src/Lecture2.hs

```diff
@@ -151,13 +168,87 @@ You're free to define any helper functions.
 -}
 
 -- some help in the beginning ;)
+
 data Knight = Knight
     { knightHealth    :: Int
     , knightAttack    :: Int
     , knightEndurance :: Int
     }
 
-dragonFight = error "TODO"
+type Gold = Int
+
+data Chest = MkChest {
+  chestGold :: Gold
+  , chestTreasure :: Maybe Treasure
+  }
+
+
+data Treasure = Sword
+              | Armor
+              | Gem
+              deriving Show
+
+data Color = Red
+           | Black
+           | Green
+            deriving Show
+
+data Dragon = Dragon
+    { color :: Color
+    , dragonHealth :: Int
+    , dragonFirePower :: Int
+    , chest :: Chest
+    }
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There're two problems with this particular data type definition:

1. It's possible to have a treasure inside green dragon:

```haskell
myDragon = Dragon
    { color = Green
    , dragonHealth = 150
    , dragonFirePower = 170
    , chest = Chest 10 (Just Sword)
    }
```

2. It's possible to not have a treasure inside non-green dragons:

```haskell
myDragon = Dragon
    { color = Green
    , dragonHealth = 150
    , dragonFirePower = 170
    , chest = Chest 10 Nothing
    }
```